In [3]:
import pandas as pd
import seaborn as sns

tips = sns.load_dataset("tips")
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

## 문자열로 변환시키기

In [8]:
tips['sex_str'] = tips['sex'].astype(str)
tips.head()

total_bill   tip     sex smoker  day    time  size sex_str
0       16.99  1.01  Female     No  Sun  Dinner     2  Female
1       10.34  1.66    Male     No  Sun  Dinner     3    Male
2       21.01  3.50    Male     No  Sun  Dinner     3    Male
3       23.68  3.31    Male     No  Sun  Dinner     2    Male
4       24.59  3.61  Female     No  Sun  Dinner     4  Female

In [7]:
tips.dtypes

total_bill     float64
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
sex_str         object
dtype: object

## 문자열로 변환된것을 다시 원래대로 돌리기

In [10]:
tips['total_bill'] = tips['total_bill'].astype(str)
tips.dtypes

total_bill      object
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
sex_str         object
dtype: object

In [12]:
tips['total_bill'] = tips['total_bill'].astype(float)
tips.dtypes

total_bill     float64
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
sex_str         object
dtype: object

# 잘못 입력한 데이터 처리하기
### 만약 정수가 있어야 하는 열에 문자열이 입력되어 있으면 어떻게 해야하나>

In [17]:
tips_sub_miss = tips.head(10)
tips_sub_miss.loc[[1, 3, 5, 7,], 'total_bill'] = 'missing'
print(tips_sub_miss)

  total_bill   tip     sex smoker  day    time  size sex_str
0      16.99  1.01  Female     No  Sun  Dinner     2  Female
1    missing  1.66    Male     No  Sun  Dinner     3    Male
2      21.01  3.50    Male     No  Sun  Dinner     3    Male
3    missing  3.31    Male     No  Sun  Dinner     2    Male
4      24.59  3.61  Female     No  Sun  Dinner     4  Female
5    missing  4.71    Male     No  Sun  Dinner     4    Male
6       8.77  2.00    Male     No  Sun  Dinner     2    Male
7    missing  3.12    Male     No  Sun  Dinner     4    Male
8      15.04  1.96    Male     No  Sun  Dinner     2    Male
9      14.78  3.23    Male     No  Sun  Dinner     2    Male


In [19]:
print(tips_sub_miss.dtypes) 

total_bill      object
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
sex_str         object
dtype: object


### 자료형을 확인해 보면 total_bill  열이 실수가 아니라 문자열임을 알 수 있다.
astype 메서드를 사용하면 변환 불가 -> 판다스는 실수로 변환하는 방법을 모르기 때문

In [21]:
tips_sub_miss['total_bill'].astype(float)

ValueError: could not convert string to float: 'missing'

In [23]:
pd.to_numeric(tips_sub_miss['total_bill'])

ValueError: Unable to parse string "missing" at position 1

### 사실 to_numeric 메서드르 사용해도 문자열을 실수로 변환할 수 없다. 하지만 numeric 메서드는 errors 인자에 raise,coerce,ignore를 지정하여 오류를 어느 정도 제어 가능

In [25]:
to_numeric_text = pd.DataFrame({
    'method' : ['raise', 'coerce', 'ignore'],
    'content': ['숫자로 변환할 수 없는 값이 있으면 오류 발생', '숫자로 변환할 수 없는 값을 누락값으로 지정',
               '아무 작업도 하지 않음']
})
to_numeric_text

method                    content
0   raise  숫자로 변환할 수 없는 값이 있으면 오류 발생
1  coerce   숫자로 변환할 수 없는 값을 누락값으로 지정
2  ignore               아무 작업도 하지 않음

### errors 인자를 ignore로 설정하면 오류가 발생하지 않지만 자료형도 변하지 않음

In [28]:
tips_sub_miss['total_bill'] = pd.to_numeric(tips_sub_miss['total_bill'],errors='ignore')
tips_sub_miss.dtypes


C:\Users\B0103\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


total_bill      object
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
sex_str         object
dtype: object

### errors 인자를 coerce(억압하다, 강요하다) 로 설정하면 missing 이 누락값(NAN) 으로 바뀜

In [29]:
tips_sub_miss['total_bill'] = pd.to_numeric(tips_sub_miss['total_bill'], errors='coerce')
tips_sub_miss.dtypes

C:\Users\B0103\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


total_bill     float64
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
sex_str         object
dtype: object

### to_numeric 매소드에서 erroes 인자 외에도 downcast 인자가 있다. downcast는 정수, 실수와 같은 자료형을 더 작은 형태로 만들 떄 사용

In [31]:
tips_sub_miss['total_bill'] = pd.to_numeric(tips_sub_miss['total_bill'], errors='coerce', downcast = 'float')
tips_sub_miss.dtypes

C:\Users\B0103\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


total_bill     float32
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
sex_str         object
dtype: object

# 카테고리 자료형

### 카테고리 자료형의 장점과 특징
- 용량과 속도 면에서 매우 효율적이다.
- 주로 동일한 문자열이 반복되어 데이터를 구성하는 경우에 사용한다.

#### 문자열을 카테고리로 변환하기

In [34]:
tips['sex'] = tips['sex'].astype('str')
tips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 8 columns):
total_bill    244 non-null float64
tip           244 non-null float64
sex           244 non-null object
smoker        244 non-null category
day           244 non-null category
time          244 non-null category
size          244 non-null int64
sex_str       244 non-null object
dtypes: category(3), float64(2), int64(1), object(2)
memory usage: 10.7+ KB


###  'str'로 변환 헀을 떄 위의 자료의 메모리 사용양은 10.7 이고  아래의 'category' 로 변환 했을떄  9.1 이다

In [36]:


tips['sex'] = tips['sex'].astype('category')
tips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 8 columns):
total_bill    244 non-null float64
tip           244 non-null float64
sex           244 non-null category
smoker        244 non-null category
day           244 non-null category
time          244 non-null category
size          244 non-null int64
sex_str       244 non-null object
dtypes: category(4), float64(2), int64(1), object(1)
memory usage: 9.2+ KB
